drivers/serial/uart_infineon_cat1.c

```diff
@@ -0,0 +1,506 @@
+#define UART_RTS_PIN  (CAT1_PIN_SIGNAL_SCB_UART_RTS)    /* RTS pin */
+#define UART_CTS_PIN  (CAT1_PIN_SIGNAL_SCB_UART_CTS)    /* CTS pin */
+
+/* Data structure */
+struct uart_cat1_data {
+	cyhal_uart_t obj;                          /* UART CYHAL object */
+	struct uart_config cfg;
+
+	#if CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t irq_cb;           /* Interrupt Callback */
+	void *irq_cb_data;                              /* Interrupt Callback Arg */
+	#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+	#ifdef CONFIG_UART_ASYNC_API
+	uart_callback_t async_cb;
+	void *async_cb_data;
+	void *async_rx_next_buf;
+	size_t async_rx_next_buf_len;
+	bool async_rx_enabled;
+	#endif /* CONFIG_UART_ASYNC_API */
+};
+
+/* Device config structure */
+struct uart_cat1_config {
+	const struct pinctrl_dev_config *pcfg;
+	struct uart_config z_cfg;
+};
+
+#define DEV_DATA(dev) \
+	((struct uart_cat1_data *const)(dev)->data)
+
+#define DEV_CFG(dev) \
+	((struct uart_cat1_config *const)(dev)->config)
+
+/* Internal function to get cyhal_gpio_t from pinctrl array by defined index. */
+extern cyhal_gpio_t cat1_pinctrl_get_cyhal_gpio(const struct pinctrl_dev_config *config,
+						uint8_t signal_name);
+
+/* Helper API */
+static cyhal_uart_parity_t _convert_uart_parity_z_to_cyhal(uint8_t z_parity)
+{
+	cyhal_uart_parity_t parity;
+
+	switch ((enum uart_config_parity) z_parity) {
+	case UART_CFG_PARITY_NONE:
+		parity = CYHAL_UART_PARITY_NONE;
+		break;
+	case UART_CFG_PARITY_ODD:
+		parity = CYHAL_UART_PARITY_ODD;
+		break;
+	case UART_CFG_PARITY_EVEN:
+		parity = CYHAL_UART_PARITY_EVEN;
+		break;
+	default:
+		parity = CYHAL_UART_PARITY_NONE;
+	}
+	return (parity);
+}
+
+static uint32_t _convert_uart_stop_bits_z_to_cyhal(uint8_t z_stop_bits)
+{
+	uint32_t stop_bits;
+
+	switch ((enum uart_config_stop_bits) z_stop_bits) {
+	case UART_CFG_STOP_BITS_1:
+		stop_bits = 1u;
+		break;
+
+	case UART_CFG_STOP_BITS_2:
+		stop_bits = 2u;
+		break;
+	default:
+		stop_bits = 1u;
+	}
+	return (stop_bits);
+}
+
+static uint32_t _convert_uart_data_bits_z_to_cyhal(uint8_t z_data_bits)
+{
+	uint32_t data_bits;
+
+	switch ((enum uart_config_data_bits) z_data_bits) {
+	case UART_CFG_DATA_BITS_5:
+		data_bits = 1u;
+		break;
+
+	case UART_CFG_DATA_BITS_6:
+		data_bits = 6u;
+		break;
+
+	case UART_CFG_DATA_BITS_7:
+		data_bits = 7u;
+		break;
+
+	case UART_CFG_DATA_BITS_8:
+		data_bits = 8u;
+		break;
+
+	case UART_CFG_DATA_BITS_9:
+		data_bits = 9u;
+		break;
+
+	default:
+		data_bits = 1u;
+	}
+	return (data_bits);
+}
+
+
+/* UART CONSOLE AND CONFIGURATION API */
+static int uart_cat1_poll_in(const struct device *dev, unsigned char *c)
+{
+	cy_rslt_t rec;
+
+	rec = cyhal_uart_getc(&DEV_DATA(dev)->obj, c, 0u);
+
+	return ((rec == CY_SCB_UART_RX_NO_DATA) ? -1 : 0);
+}
+
+static void uart_cat1_poll_out(const struct device *dev, unsigned char c)
+{
+	(void) cyhal_uart_putc(&DEV_DATA(dev)->obj, (uint32_t)c);
+}
```

Comment:
What happen if buffer is full? Is this blocking?

Reply:
https://github.com/zephyrproject-rtos/zephyr/pull/44211#issuecomment-1202399515